metadata/src/metadata/index.js

```diff
@@ -0,0 +1,79 @@
+const ts = require('typescript');
+const fs = require('fs');
+const path = require('path');
+const generateModuleMetadata = require('./module-metadata');
+const palletsPath = path.join(__dirname, "../../../assembly/pallets/")
+
+/**
+ * Generates the metadata of the Subsembly runtime
+ */
+function generateMetadata() {
+    /**
+     * Template metadata object to be populated
+     */
+    const metadata = {
+        magicNumber: 1635018093,
+        metadata: {
+            V12: {
+                modules: [],
+                extrinsic: {
+                    version: 4,
+                    signedExtensions: [
+                        "CheckSpecVersion",
+                        "CheckTxVersion",
+                        "CheckGenesis",
+                        "CheckMortality",
+                        "CheckNonce",
+                        "CheckWeight",
+                        "ChargeTransactionPayment"
+                    ]
+                }
+            }
+        }
+    };
+
+    /**
+     * System module is not inside pallets directory, so we load it separately
+     */
+    const systemNode = ts.createSourceFile(
+        "System",
+        fs.readFileSync(path.join(__dirname, '../../../assembly/frame/system.ts'), 'utf-8'),
+        ts.ScriptTarget.Latest
+    );
+
+    /**
+     * Add System module metadata
+     */
+    metadata.metadata.V12.modules.push(generateModuleMetadata(metadata.metadata.V12.modules.length, systemNode));
+
+    /**
+     * Go through the pallets inside `pallets` folder and generate module metadata for each pallet
+     */
+    fs.readdirSync(palletsPath).forEach(module => {
+        // path to the top AS project folder
+        const assemblyPath = path.join(palletsPath, module, "assembly");
+        // we ignore index.ts and json files
+        if (!module.includes("index.ts") && !module.includes("json")) {
+            // Go through the files inside the pallet
+            fs.readdirSync(assemblyPath).forEach(file => {
+                // Name of the file should be same as the name of the pallet/module
+                if (file.includes(module)) {
+                    const moduleNode = ts.createSourceFile(
+                        module,
+                        fs.readFileSync(path.join(assemblyPath, file), "utf-8"),
+                        ts.ScriptTarget.Latest
+                    );
+                    metadata.metadata.V12.modules.push(generateModuleMetadata(metadata.metadata.V12.modules.length, moduleNode));
+                }
+            });
+        }
+    }
+    );
+
+    /**
+     * Write to the json file in parent directory
+     */
+    fs.writeFileSync("../../metadata.json", JSON.stringify(metadata, null, 4));
```

Comment:
I think that instead of creating the metadata.json file (for debugging purposes we could - f.e add additional property `-debug`  - if set to `true` we could save the **non hex encoded**  version of the metadata the way you do here, **however**, imo the production version of the tool should be:

1) generate the metadata.json (as you do now)
2) convert the metadata.json to hex encoded byte array
3) Place the hex encoded byte array to the `metadata.ts` file in the `./assembly/frame/metadata.ts`

I guess that we can do that in a separate PR as-well (step 3 + stitching together all steps in one npm command and adding it before compilation is executed)